### ISSUE TITLE
GroupingSet::toIntermediate() clears Aggregate::numNull_ flag before destroying accumulators

### ISSUE THREAD

**Issue by kagamiori** [facebookincubator/velox#5823]
### Bug description

When a UDAF doesn't have its own toIntermediate() implementation, GroupingSet::toIntermediate() calls function->addRawInput() and function->extractAccumulators() to compute the intermediate result, and then calls RowContainer::eraseRows() to free the RowContainer. eraseRows() calls Aggregate::destroy() which, for some UDAFs that use ValueList, should frees up the accumulator only at non-null groups (through ValueList::free()). However, GroupingSet::toIntermediate() currently calls Aggregate::clear() that resets Aggregate::numNulls_ before calling eraseRows(). This causes Aggregate::destroy() to consider all groups to be non-null and attempt to free non-existing accumulators at null-groups. This caused SIGSEGV on the test of https://github.com/facebookincubator/velox/pull/5643 and is hence blocking https://github.com/facebookincubator/velox/pull/5643 from landing.

### System information

N/A

### Relevant logs

```bash
*** Aborted at 1690304155 (Unix time, try 'date -d @1690304155') ***
*** Signal 11 (SIGSEGV) (0x0) received by PID 2751637 (pthread TID 0x7fdedec00640) (linux TID 2753308) (code: 128), stack trace: ***
    @ 0000000000843773 (unknown)
    @ 0000000000842c2d (unknown)
    @ 000000000004459f (unknown)
    @ 000000000007908c facebook::velox::HashStringAllocator::Header::size() const
                       ./fbcode/velox/common/memory/HashStringAllocator.h:110
                       -> ./fbcode/velox/common/memory/HashStringAllocator.cpp
    @ 0000000000079620 facebook::velox::HashStringAllocator::free(facebook::velox::HashStringAllocator::Header*)
                       ./fbcode/velox/common/memory/HashStringAllocator.cpp:421
    @ 00000000007734a1 (unknown)
    @ 000000000077344c (unknown)
    @ 000000000076b8ad (unknown)
    @ 00000000005957f8 facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0::operator()(folly::Range<char**>) const
                       ./fbcode/velox/exec/RowContainer.cpp:56
    @ 00000000005957a5 void std::__invoke_impl<void, facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0&, folly::Range<char**> >(std::__invoke_other, facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0&, folly::Range<char**>&&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/invoke.h:61
                       -> ./fbcode/velox/exec/RowContainer.cpp
    @ 0000000000595721 std::enable_if<is_invocable_r_v<void, facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0&, folly::Range<char**> >, void>::type std::__invoke_r<void, facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0&, folly::Range<char**> >(facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0&, folly::Range<char**>&&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/invoke.h:111
                       -> ./fbcode/velox/exec/RowContainer.cpp
    @ 0000000000595611 std::_Function_handler<void (folly::Range<char**>), facebook::velox::exec::Accumulator::Accumulator(facebook::velox::exec::Aggregate*)::$_0>::_M_invoke(std::_Any_data const&, folly::Range<char**>&&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/std_function.h:291
                       -> ./fbcode/velox/exec/RowContainer.cpp
    @ 0000000000592388 std::function<void (folly::Range<char**>)>::operator()(folly::Range<char**>) const
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/std_function.h:560
                       -> ./fbcode/velox/exec/RowContainer.cpp
    @ 0000000000592328 facebook::velox::exec::Accumulator::destroy(folly::Range<char**>)
                       ./fbcode/velox/exec/RowContainer.cpp:91
    @ 0000000000593ff3 facebook::velox::exec::RowContainer::freeAggregates(folly::Range<char**>)
                       ./fbcode/velox/exec/RowContainer.cpp:355
    @ 0000000000594069 facebook::velox::exec::RowContainer::eraseRows(folly::Range<char**>)
                       ./fbcode/velox/exec/RowContainer.cpp:288
    @ 000000000044c364 facebook::velox::exec::GroupingSet::toIntermediate(std::shared_ptr<facebook::velox::RowVector> const&, std::shared_ptr<facebook::velox::RowVector>&)
                       ./fbcode/velox/exec/GroupingSet.cpp:1065
    @ 0000000000454fdf facebook::velox::exec::HashAggregation::getOutput()
                       ./fbcode/velox/exec/HashAggregation.cpp:359
    @ 00000000003bf399 facebook::velox::exec::Driver::runInternal(std::shared_ptr<facebook::velox::exec::Driver>&, std::shared_ptr<facebook::velox::exec::BlockingState>&, std::shared_ptr<facebook::velox::RowVector>&)
                       ./fbcode/velox/exec/Driver.cpp:425
    @ 00000000003c1f6b facebook::velox::exec::Driver::run(std::shared_ptr<facebook::velox::exec::Driver>)
                       ./fbcode/velox/exec/Driver.cpp:554
    @ 00000000003c7e15 facebook::velox::exec::Driver::enqueue(std::shared_ptr<facebook::velox::exec::Driver>)::$_1::operator()() const
                       ./fbcode/velox/exec/Driver.cpp:206
    @ 00000000003c7c5c void folly::detail::function::FunctionTraits<void ()>::callSmall<facebook::velox::exec::Driver::enqueue(std::shared_ptr<facebook::velox::exec::Driver>)::$_1>(folly::detail::function::Data&)
                       ./fbcode/folly/Function.h:345
                       -> ./fbcode/velox/exec/Driver.cpp
    @ 0000000000029fe1 folly::detail::function::FunctionTraits<void ()>::operator()()
                       ./fbcode/folly/Function.h:374
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 0000000000029309 folly::ThreadPoolExecutor::runTask(std::shared_ptr<folly::ThreadPoolExecutor::Thread> const&, folly::ThreadPoolExecutor::Task&&)
                       ./fbcode/folly/lang/Exception.h:285
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 0000000000041997 folly::CPUThreadPoolExecutor::threadRun(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)
                       ./fbcode/folly/executors/CPUThreadPoolExecutor.cpp:333
    @ 000000000002cba6 void std::__invoke_impl<void, void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>(std::__invoke_memfun_deref, void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/invoke.h:74
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 000000000002ca66 std::__invoke_result<void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>::type std::__invoke<void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&>(void (folly::ThreadPoolExecutor::*&)(std::shared_ptr<folly::ThreadPoolExecutor::Thread>), folly::ThreadPoolExecutor*&, std::shared_ptr<folly::ThreadPoolExecutor::Thread>&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/invoke.h:96
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 000000000002ca01 void std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>::__call<void, , 0ul, 1ul>(std::tuple<>&&, std::_Index_tuple<0ul, 1ul>)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/functional:420
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 000000000002c995 void std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)>::operator()<, void>()
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/functional:503
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 000000000002c72c void folly::detail::function::FunctionTraits<void ()>::callSmall<std::_Bind<void (folly::ThreadPoolExecutor::*(folly::ThreadPoolExecutor*, std::shared_ptr<folly::ThreadPoolExecutor::Thread>))(std::shared_ptr<folly::ThreadPoolExecutor::Thread>)> >(folly::detail::function::Data&)
                       ./fbcode/folly/Function.h:345
                       -> ./fbcode/folly/executors/ThreadPoolExecutor.cpp
    @ 00000000001937f1 folly::detail::function::FunctionTraits<void ()>::operator()()
                       ./fbcode/folly/Function.h:374
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 0000000000193766 folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}::operator()()
                       ./fbcode/folly/executors/thread_factory/NamedThreadFactory.h:40
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 000000000019371c void std::__invoke_impl<void, folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>(std::__invoke_other, folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/invoke.h:61
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 00000000001936ac std::__invoke_result<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>::type std::__invoke<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}>(folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}&&)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/invoke.h:96
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 0000000000193684 void std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::_M_invoke<0ul>(std::_Index_tuple<0ul>)
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/std_thread.h:253
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 0000000000193654 std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> >::operator()()
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/std_thread.h:260
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 00000000001934f8 std::thread::_State_impl<std::thread::_Invoker<std::tuple<folly::NamedThreadFactory::newThread(folly::Function<void ()>&&)::{lambda()#1}> > >::_M_run()
                       ././fbcode/third-party-buck/platform010/build/libgcc/include/c++/trunk/bits/std_thread.h:211
                       -> ./fbcode/velox/exec/tests/utils/OperatorTestBase.cpp
    @ 00000000000df4e4 execute_native_thread_routine
    @ 000000000009ac0e start_thread
    @ 000000000012d1db __clone3
```

**Comment by kagamiori** [comment]
Here is a fix: [D47746429](https://www.internalfb.com/diff/D47746429).

**Comment by mbasmanova** [comment]
CC: @Yuhta

**Comment by kagamiori** [comment]
I searched the code base and I saw that Aggregate::clear() is only called in GroupingSet::toIntermediate() and in AggregateWindow.cpp. Since clear() resets Aggregate::numNulls_, it should be called in a situation where the aggregation work for the current input is complete and the aggregation function instance will be reused to process another input. The use case in AggregateWindow.cpp fall in this bucket. On the other hand, GroupingSet::toIntermediate() is called at the end of HashAggregation::getOutput(), hence the aggregation function instance is unlikely to be used later after it produces the result. So I think calling Aggregate::clear() at the end of GroupingSet::toIntermediate() is not necessary.

Also, I noticed that AggregateWindowFunction::simpleAggregation() only calls Aggregate::clear() but doesn't call Aggregate::destroy(). This may possibly cause memory leak during the aggregation.
